lib/rust/ensogl/core/src/display/camera/camera2d.rs

```diff
@@ -425,7 +431,19 @@ impl Camera2d {
 
     /// Update all dirty camera parameters and compute updated view-projection matrix.
     pub fn update(&self, scene: &Scene) -> bool {
-        self.data.borrow_mut().update(scene)
+        let (is_updated, pos_endpoint, zoom_endpoint, zoom) = {
+            let mut borrowed = self.data.borrow_mut();
+            let is_updated = borrowed.update(scene);
+            let pos_endpoint = borrowed.frp.position.clone_ref();
+            let zoom_endpoint = borrowed.frp.zoom.clone_ref();
+            let zoom = borrowed.zoom;
+            (is_updated, pos_endpoint, zoom_endpoint, zoom)
+        };
+        if is_updated {
+            pos_endpoint.emit(self.display_object.position());
+            zoom_endpoint.emit(zoom);
+        }
+        is_updated
```

Comment:
1. This code duplicates the code in `CameraData`. There is no explanation why. Is the code in `CameraData` still used? If no, why its still there. If yes, why?
2. This code uses a lot of strange clone_refs inside - there is not explanation why.
3. What does "_endpoint" suffix mean? Does it mean the FRP endpoint? If so, this is like naming a variable "age_var".
4. There is something bad with this code in general. It is too complex. It should be refactored somehow, but now, it does a lot of operations that are not obvious to the reader.

Reply:
1. Line 436
2. See point 4.
3. But you see the name clash, don't you? Anyway, I disagree: It's not like naming the variable `_var`. It's more like suffixing with `_value` - sometimes it's obvious that the var keeps value, but sometimes it's a ref, a cell, or, well, an FRP endpoint.
4. It was for my request: we don't want to emit any events having anything borrowed. Thus, we must clone_ref the FRP endpoints. I don't see any simpler way for doing that, except repeating the borrow() multiple times.

Reply:
I simplified the code by copying the `frp` structure, not individual fields.

Reply:
1. Oh, there it is!
2. -
3. The name clash doesn't matter here really. I was referring to var names. If there is a name clash, between var `age` and `age`, renaming it to `age_var` is not a good solution. So I don't think the name clash should really contribute to this point. However, the explanation with "value-based suffixes" is understandable and I was not looking at it this way. I still somehow don't like it, but with this explanation, it makes much more sense in my head now.
4. I was never told that we should not do it. The request to never emit events from borrowed code is 100% valid. I was originally referring to the complexity of the implementation in general. The new version is way cleaner. If we did not want to clone the whole frp network, we could instead create a method on it allowing us to clone a sub-structure of it. However, if this is the frp network created with the new macro, we can just clone the input or output ports instead (they all are enclosed in a single RC). Anyway, @farmaazon, I was referring to code complexity, not to logical complexity.